ext/standard/streamsfuncs.c

```diff
@@ -756,115 +810,160 @@ static int stream_array_emulate_read_fd_set(zval *stream_array)
-		}
+	if (max_set_count == 0) {
+		RETURN_FALSE;
 	}
 
-	/* If seconds is not set to null, build the timeval, else we wait indefinitely */
-	if (!secnull) {
-		if (sec < 0) {
-			zend_argument_value_error(4, "must be greater than or equal to 0");
-			RETURN_THROWS();
-		} else if (usec < 0) {
-			zend_argument_value_error(5, "must be greater than or equal to 0");
-			RETURN_THROWS();
-		}
+    if (secnull && !usecnull && usec != 0) {
+		zend_argument_value_error(5, "must be null when argument #4 ($seconds) is null");
+		FD_BIGSET_FREE(&rfds);
+		FD_BIGSET_FREE(&wfds);
+		FD_BIGSET_FREE(&efds);
+		RETURN_THROWS();
+    }
 
-		/* Windows, Solaris and BSD do not like microsecond values which are >= 1 sec */
-		tv.tv_sec = (long)(sec + (usec / 1000000));
-		tv.tv_usec = (long)(usec % 1000000);
-		tv_p = &tv;
-	}
+    if (!secnull) {
+        if (sec < 0) {
+            zend_argument_value_error(4, "must be greater than or equal to 0");
+            FD_BIGSET_FREE(&rfds);
+            FD_BIGSET_FREE(&wfds);
+            FD_BIGSET_FREE(&efds);
+            RETURN_THROWS();
+        } else if (usec < 0) {
+            zend_argument_value_error(5, "must be greater than or equal to 0");
+            FD_BIGSET_FREE(&rfds);
+            FD_BIGSET_FREE(&wfds);
+            FD_BIGSET_FREE(&efds);
+            RETURN_THROWS();
+        }
+
+        tv.tv_sec = (long)(sec + (usec / 1000000));
+        tv.tv_usec = (long)(usec % 1000000);
+        tv_p = &tv;
+    }
 
 	/* slight hack to support buffered data; if there is data sitting in the
 	 * read buffer of any of the streams in the read array, let's pretend
 	 * that we selected, but return only the readable sockets */
 	if (r_array != NULL) {
-		retval = stream_array_emulate_read_fd_set(r_array);
-		if (retval > 0) {
-			if (w_array != NULL) {
-				zval_ptr_dtor(w_array);
-				ZVAL_EMPTY_ARRAY(w_array);
-			}
-			if (e_array != NULL) {
-				zval_ptr_dtor(e_array);
-				ZVAL_EMPTY_ARRAY(e_array);
-			}
-			RETURN_LONG(retval);
-		}
-	}
+        retval = stream_array_emulate_read_fd_set(r_array);
+        if (retval > 0) {
+            if (w_array != NULL) {
+                zval_ptr_dtor(w_array);
+                ZVAL_EMPTY_ARRAY(w_array);
+            }
+            if (e_array != NULL) {
+                zval_ptr_dtor(e_array);
+                ZVAL_EMPTY_ARRAY(e_array);
+            }
+            FD_BIGSET_FREE(&rfds);
+            FD_BIGSET_FREE(&wfds);
+            FD_BIGSET_FREE(&efds);
+            RETURN_LONG(retval);
+        }
+    }
 
-	retval = php_select(max_fd+1, &rfds, &wfds, &efds, tv_p);
 
-	if (retval == -1) {
-		php_error_docref(NULL, E_WARNING, "Unable to select [%d]: %s (max_fd=%d)",
-				errno, strerror(errno), max_fd);
-		RETURN_FALSE;
-	}
+    /* Ensure all bitmaps are the same size (one or more may have been dynamically resized) */
+    size_t largest = rfds.size;
+    if (largest < wfds.size) largest = wfds.size;
+    if (largest < efds.size) largest = efds.size;
```

Comment:
I think we can use `max_fd` to directly compute the largest size?

Reply:
Something like:

	while (max_fd > rfds.size) fd_bigset_double_size(&rfds);

I was just not confident about possibly off by one error here...

Reply:
We can use the same condition as in FD_BIGSET_ENSURE_CAPACITY(), or pre-compute the largest size with:

```
size_t num_fds = max_fd + 1;
size_t largest = (num_fds + 7) / 8;
```